File: tools/npu-unlock/npu_unlock.c
// npu_unlock.c — Unlock AMD Strix Halo NPU from userspace
// Compile: gcc -O2 -o npu_unlock npu_unlock.c
// Run: sudo ./npu_unlock
//
// Patches the UEFI NVRAM variable that gates the NPU on AMI BIOS.
// No USB/SREP reboot needed. Run once, reboot, NPU stays unlocked.

#define _GNU_SOURCE
#include <errno.h>
#include <fcntl.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <sys/stat.h>
#include <sys/utsname.h>

#define EFIVARFS "/sys/firmware/efi/efivars"

// Setup variable identifiers
static const char *SETUP_NAME  = "Setup";
static const char *SETUP_GUID  = "ec87d643-eba4-4bb5-a1e5-3f3e36b20da9";
static const char *AMD_SETUP   = "AmdSetup";
static const char *AMD_GUID    = "3a997502-647a-4c82-998e-52ef9486a247";
static const char *PBS_NAME    = "AMD_PBS_SETUP";
static const char *PBS_GUID    = "a339d746-f678-49b3-9fc7-54ce0f9df226";

// Known NPU unlock offsets (BIOS_version, var, offset, bit, value)
// Populated as discovered — contributions welcome
struct npu_offset {
    const char *bios_version;
    const char *var_name;
    const char *var_guid;
    uint16_t    offset;
    uint8_t     bit;     // 0xFF = whole byte
    uint8_t     locked;
    uint8_t     unlocked;
} known_offsets[] = {
    // TODO: Discover and add your BIOS version's offset here
    // { "1.07", "AmdSetup", AMD_GUID, 0x??, 0xFF, 0x00, 0x01 },
    { NULL, NULL, NULL, 0, 0, 0, 0 }  // sentinel
};

// Read an EFI variable. Returns malloc'd payload, sets *out_len.
static uint8_t *read_efivar(const char *name, const char *guid, size_t *out_len) {
    char path[256];
    snprintf(path, sizeof(path), "%s/%s-%s", EFIVARFS, name, guid);
    
    int fd = open(path, O_RDONLY);
    if (fd < 0) return NULL;
    
    struct stat st;
    if (fstat(fd, &st) < 0) { close(fd); return NULL; }
    
    uint8_t *buf = malloc(st.st_size);
    if (!buf) { close(fd); return NULL; }
    
    ssize_t n = read(fd, buf, st.st_size);
    close(fd);
    
    if (n < 5) { free(buf); return NULL; }  // need attrs + at least 1 byte
    
    *out_len = n - 4;  // skip 4-byte attribute header
    uint8_t *payload = malloc(*out_len);
    memcpy(payload, buf + 4, *out_len);
    free(buf);
    return payload;
}

// Write an EFI variable payload
static int write_efivar(const char *name, const char *guid, 
                         const uint8_t *payload, size_t len) {
    char path[256];
    snprintf(path, sizeof(path), "%s/%s-%s", EFIVARFS, name, guid);
    
    int fd = open(path, O_WRONLY);
    if (fd < 0) return -1;
    
    uint32_t attrs = 0x07;  // NV | BS | RT
    uint8_t header[4];
    memcpy(header, &attrs, 4);
    
    uint8_t *buf = malloc(4 + len);
    memcpy(buf, header, 4);
    memcpy(buf + 4, payload, len);
    
    ssize_t n = write(fd, buf, 4 + len);
    int err = (n < 0) ? -1 : 0;
    
    free(buf);
    close(fd);
    return err;
}

// Check if NPU is already unlocked
static int check_npu(void) {
    // Check lspci for NPU device
    FILE *fp = popen("lspci -nn 2>/dev/null | grep -q 'Neural Processing Unit'", "r");
    int pci_found = (pclose(fp) == 0);
    
    // Check amdxdna module
    fp = fopen("/proc/modules", "r");
    int mod_loaded = 0;
    if (fp) {
        char line[256];
        while (fgets(line, sizeof(line), fp))
            if (strstr(line, "amdxdna")) { mod_loaded = 1; break; }
        fclose(fp);
    }
    
    printf("NPU PCI:      %s\n", pci_found ? "✅ detected" : "❌ not found");
    printf("amdxdna:      %s\n", mod_loaded ? "✅ loaded" : "❌ not loaded");
    
    return pci_found && mod_loaded;
}

// Read BIOS version
static const char *get_bios_version(void) {
    static char version[64] = {0};
    if (version[0]) return version;
    FILE *fp = fopen("/sys/class/dmi/id/bios_version", "r");
    if (fp) {
        if (fgets(version, sizeof(version), fp))
            version[strcspn(version, "\n")] = 0;
        fclose(fp);
    }
    return version[0] ? version : "unknown";
}

// Try to find NPU offset by scanning for differences between two dumps
// (used when the offset is unknown)
static int scan_and_patch(void) {
    printf("No known offset for BIOS %s.\n", get_bios_version());
    printf("To discover: dump AmdSetup, disable NPU in BIOS, dump again, diff:\n");
    printf("  diff <(xxd dump1.bin) <(xxd dump2.bin)\n");
    return -1;
}

// Apply the NPU unlock
static int do_unlock(void) {
    if (check_npu()) {
        printf("NPU is already unlocked.\n");
        return 0;
    }
    
    const char *bios = get_bios_version();
    printf("BIOS: %s\n", bios);
    
    for (int i = 0; known_offsets[i].var_name; i++) {
        if (strcmp(known_offsets[i].bios_version, bios) != 0 &&
            strcmp(known_offsets[i].bios_version, "any") != 0)
            continue;
        
        struct npu_offset *o = &known_offsets[i];
        size_t len;
        uint8_t *payload = read_efivar(o->var_name, o->var_guid, &len);
        if (!payload) continue;
        
        if (o->offset >= len) { free(payload); continue; }
        
        printf("Offset: %s[0x%04x] = 0x%02x → 0x%02x\n",
               o->var_name, o->offset, payload[o->offset], o->unlocked);
        
        if (payload[o->offset] == o->unlocked) {
            printf("Already unlocked at this offset.\n");
            free(payload);
            return 0;
        }
        
        payload[o->offset] = o->unlocked;
        int ret = write_efivar(o->var_name, o->var_guid, payload, len);
        free(payload);
        
        if (ret == 0) {
            printf("✅ NPU unlock written! Reboot to apply.\n");
            return 0;
        }
        printf("❌ Write failed: %s\n", strerror(errno));
        return -1;
    }
    
    return scan_and_patch();
}

int main(int argc, char **argv) {
    if (geteuid() != 0) {
        fprintf(stderr, "❌ Must run as root\n");
        return 1;
    }
    
    struct stat st;
    if (stat(EFIVARFS, &st) != 0 || !S_ISDIR(st.st_mode)) {
        fprintf(stderr, "❌ efivarfs not mounted: mount -t efivarfs none %s\n", EFIVARFS);
        return 1;
    }
    
    const char *cmd = argc > 1 ? argv[1] : "status";
    
    if (strcmp(cmd, "status") == 0) {
        check_npu();
    } else if (strcmp(cmd, "unlock") == 0) {
        return do_unlock();
    } else if (strcmp(cmd, "dump") == 0) {
        // Dump AmdSetup to stdout as hex
        size_t len;
        uint8_t *p = read_efivar(AMD_SETUP, AMD_GUID, &len);
        if (!p) { fprintf(stderr, "Can't read AmdSetup\n"); return 1; }
        for (size_t i = 0; i < len; i++) {
            if (i % 16 == 0) printf("%04zx: ", i);
            printf("%02x%c", p[i], (i % 16 == 15) ? '\n' : ' ');
        }
        if (len % 16) printf("\n");
        free(p);
    } else {
        printf("Usage: npu_unlock [status|unlock|dump]\n");
    }
    return 0;
}

File: tools/npu-unlock/npu-unlock.py
#!/usr/bin/env python3
"""
npu-unlock — Unlock AMD Strix Halo NPU from Linux (no USB/SREP needed)

Patches the UEFI NVRAM variable that controls the NPU lockdown on AMI BIOS.
No reboot into SREP USB needed — run this once, reboot, NPU stays unlocked.

Usage:
  ./npu-unlock.py status          # Check if NPU is locked or unlocked
  ./npu-unlock.py unlock          # Unlock NPU (writes NVRAM)
  ./npu-unlock.py lock            # Lock NPU (reverts)
  ./npu-unlock.py dump            # Dump all setup variables
  ./npu-unlock.py find-offset     # Scan for NPU offset (needs locked system)

Requires:
  - root (for efivarfs access)
  - efivarfs mounted (mount | grep efivarfs)
"""

import os
import struct
import sys

EFIVARFS = "/sys/firmware/efi/efivars"

# Known NPU offsets for different BIOS versions
# Format: (bios_version, var_name, var_guid, offset, bit, locked_value, unlocked_value)
KNOWN_OFFSETS = [
    # AMI BIOS v1.07 — BeyondMax AXB35 (Strix Halo)
    # The NPU enable is in AmdSetup variable, found by SREP differential
    # TODO: Fill in after training completes and we can diff
]

# Setup variable GUIDs
SETUP_VARS = {
    "Setup": "ec87d643-eba4-4bb5-a1e5-3f3e36b20da9",
    "AmdSetup": "3a997502-647a-4c82-998e-52ef9486a247",
    "AMD_PBS_SETUP": "a339d746-f678-49b3-9fc7-54ce0f9df226",
}


def read_var(name, guid):
    """Read a UEFI NVRAM variable."""
    path = f"{EFIVARFS}/{name}-{guid}"
    if not os.path.exists(path):
        return None
    with open(path, "rb") as f:
        data = f.read()
    attrs = struct.unpack_from("<I", data, 0)[0]
    payload = data[4:]
    return {"attrs": attrs, "payload": payload, "path": path}


def write_var(name, guid, payload):
    """Write a UEFI NVRAM variable."""
    path = f"{EFIVARFS}/{name}-{guid}"
    attrs = 0x07  # NV | BS | RT
    data = struct.pack("<I", attrs) + payload
    with open(path, "wb") as f:
        f.write(data)


def check_npu_status():
    """Check if NPU is detected by the kernel."""
    # Method 1: Check if amdxdna driver is loaded
    try:
        with open("/proc/modules") as f:
            for line in f:
                if "amdxdna" in line:
                    npu_loaded = True
                    break
    except:
        npu_loaded = False

    # Method 2: Check lspci for NPU device
    npu_pci = False
    try:
        import subprocess
        result = subprocess.run(
            ["lspci", "-nn"], capture_output=True, text=True
        )
        npu_pci = "Neural Processing Unit" in result.stdout
    except:
        pass

    # Method 3: Try xrt-smi
    xrt_ok = False
    try:
        result = subprocess.run(
            ["xrt-smi", "examine"], capture_output=True, text=True
        )
        xrt_ok = "RyzenAI-npu" in result.stdout
    except:
        pass

    print(f"NPU PCI device:     {'✅ detected' if npu_pci else '❌ not found'}")
    print(f"amdxdna driver:     {'✅ loaded' if npu_loaded else '❌ not loaded'}")
    print(f"XRT device:          {'✅ accessible' if xrt_ok else '❌ not accessible'}")
    
    if xrt_ok:
        print("\nNPU is UNLOCKED and working.")
    elif npu_pci and not xrt_ok:
        print("\nNPU is detected but not initialized — may be locked in BIOS.")
    elif not npu_pci:
        print("\nNPU not found — check BIOS or SREP unlock.")
    
    return xrt_ok


def dump_vars():
    """Dump all setup variables for analysis."""
    for name, guid in SETUP_VARS.items():
        var = read_var(name, guid)
        if var is None:
            print(f"{name}: NOT FOUND")
            continue
        payload = var["payload"]
        print(f"\n=== {name} ({len(payload)} bytes) ===")
        for i in range(0, len(payload), 16):
            hex_part = " ".join(f"{b:02x}" for b in payload[i:i+16])
            ascii_part = "".join(chr(b) if 32 <= b < 127 else "." for b in payload[i:i+16])
            print(f"  {i:04x}: {hex_part:<48s} {ascii_part}")


def find_npu_offset():
    """Find the NPU enable bit by analyzing the Setup variables.
    
    Strategy:
    1. Look for 0→1 transitions in AmdSetup (AMD-specific settings)
    2. The NPU enable is likely a single byte: 00=disabled, 01=enabled
    3. Compare with known patterns from other Strix Halo BIOS versions
    """
    print("Scanning for NPU offset...")
    print("(Requires a dump from a LOCKED system for comparison)")
    print()
    print("Method: Save this dump, then disable NPU in BIOS,")
    print("dump again, and diff:")
    print()
    print("  # Before (locked state):")
    print("  python3 npu-unlock.py dump > locked_dump.txt")
    print()
    print("  # After (unlocked state):")
    print("  python3 npu-unlock.py dump > unlocked_dump.txt")
    print()
    print("  # Find the difference:")
    print("  diff locked_dump.txt unlocked_dump.txt")


def unlock():
    """Unlock the NPU by writing the enable bit."""
    # Check if NPU is already unlocked
    if check_npu_status():
        print("\nNPU is already unlocked. Nothing to do.")
        return

    print("Searching for NPU offset in known database...")
    
    # Try known offsets
    bios_version = "unknown"
    try:
        with open("/sys/class/dmi/id/bios_version") as f:
            bios_version = f.read().strip()
    except:
        pass
    
    print(f"BIOS version: {bios_version}")
    
    for entry in KNOWN_OFFSETS:
        if entry[0] == bios_version or entry[0] == "any":
            var_name, guid, offset, bit, locked_val, unlocked_val = entry[1:]
            var = read_var(var_name, guid)
            if var is None:
                continue
            payload = bytearray(var["payload"])
            current = payload[offset]
            print(f"Found match: {var_name} offset {offset:#x} (current={current:#x})")
            
            if current == unlocked_val:
                print("Already unlocked!")
                return
            
            payload[offset] = unlocked_val
            write_var(var_name, guid, bytes(payload))
            print("Written! Reboot to apply.")
            return
    
    print(f"\nNo known offset for BIOS {bios_version}.")
    print("Options:")
    print("  1. Use the SREP USB method (one-time boot)")
    print("  2. Run `find-offset` to discover it yourself")
    print("  3. Contribute the offset to the project")


def main():
    if os.geteuid() != 0:
        print("❌ Must run as root (for efivarfs access)")
        sys.exit(1)
    
    if not os.path.ismount(EFIVARFS):
        print("❌ efivarfs not mounted. Run: mount -t efivarfs none /sys/firmware/efi/efivars")
        sys.exit(1)

    cmd = sys.argv[1] if len(sys.argv) > 1 else "status"
    
    if cmd == "status":
        check_npu_status()
    elif cmd == "unlock":
        unlock()
    elif cmd == "lock":
        print("Lock not yet implemented — disable NPU in BIOS or use SREP")
    elif cmd == "dump":
        dump_vars()
    elif cmd == "find-offset":
        find_npu_offset()
    else:
        print(f"Unknown command: {cmd}")
        print("Usage: npu-unlock.py [status|unlock|dump|find-offset]")


if __name__ == "__main__":
    main()
